### PR TITLE
Get a chromosome using chromosome's name along with (assembly's tax id or assembly)

### PR DESCRIPTION
**Pull request by singaltanmay**

### CODE REVIEW

**Comment by singaltanmay** [comment]
image
There are assemblies that have identical chromosomes associated to them. Should I filter the result of the query and remove duplicates or should I let them stay? Keeping them can make sense if we want to link parent assemblies using HateOAS links because identical chromosomes would point to different assemblies.

**Comment by singaltanmay** [comment]
Embed the assembly object in the chromosome response (Not HateOAS links).

**Comment by singaltanmay** [comment]
> Embed the assembly object in the chromosome response (Not HateOAS links).

Done

File: src/test/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasHandlerTest.java [hunk 1]
```diff
@@ -230,21 +231,54 @@ void assertChromosomeIdenticalToEntity(ChromosomeEntity chromosome) {
...
+            Mockito.when(
+                    mockChromosomeService.getChromosomesByNameAndAssemblyTaxid(chrName, asmTaxid, DEFAULT_PAGE_REQUEST))
+                   .thenReturn(new PageImpl<>(chromosomeEntities
+                                                      .parallelStream()
+                                                      .filter(it -> it.getName().equals(chrName) &&
+                                                              it.getAssembly().getTaxid().equals(asmTaxid))
+                                                      .collect(Collectors.toList())));
+
+            List<ChromosomeEntity> chromosomesByNameAndAssembly = chromosomeEntities
+                    .parallelStream()
```
Comment (jmmut) [inline src/test/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasHandlerTest.java:260]: why parallel stream? speed? have you measured it?

File: src/test/java/uk/ac/ebi/eva/contigalias/service/AssemblyAndChromosomeServiceIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        void tearDown() {
+            for (AssemblyEntity assemblyEntity : assemblyEntities) {
+                assemblyService.deleteAssembly(assemblyEntity);
+            }
+            assemblyEntities = null;
+            chromosomeEntities = null;
+        }
+
+        /**
+         * In this test two chromosomes have non-distinct names and are linked to different assemblies except that those
```
Comment (jmmut) [inline src/test/java/uk/ac/ebi/eva/contigalias/service/AssemblyAndChromosomeServiceIntegrationTest.java:187]: I'm not sure why you say "non-distinct" instead of "the same". Is there any detail you intended to highlight?

File: src/main/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasController.java [hunk 3]
```diff
@@ -169,23 +169,58 @@ public ContigAliasController(ContigAliasHandler handler) {
...
-        } else {
-            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
+        return createAppropriateResponseEntity(entities);
+    }
+
+    @ApiOperation(value = "Get chromosomes using a combination of their own name and the Taxonomic ID's of their " +
+            "parent assemblies.",
+            notes = "Given a chromosome's name and the Taxonomic ID of the assembly that it belongs to, this endpoint" +
+                    " will return a list of chromosomes that satisfy the given parameters. Each chromosome will also " +
+                    "have it's parent assembly nested inside it. The endpoint will either return a list of " +
```
[the remark this answers is not in the document]
Comment (singaltanmay) [reply src/main/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasController.java:186]: I really need to stop making this mistake.

File: src/main/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasController.java [hunk 4]
```diff
@@ -169,23 +169,58 @@ public ContigAliasController(ContigAliasHandler handler) {
...
+                    "chromosomes or it will either return an HTTP error code 204 if no chromosomes are found or " +
```
[the remark this answers is not in the document]
Comment (singaltanmay) [reply src/main/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasController.java:187]: I have changed all to return 204.

File: src/test/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasHandlerTest.java (hunk 1 above)
Comment (singaltanmay) [reply src/test/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasHandlerTest.java:260]: I haven't measured it. Parallelized it for no good reason. Have removed it now.

File: src/test/java/uk/ac/ebi/eva/contigalias/service/AssemblyAndChromosomeServiceIntegrationTest.java (hunk 2 above)
Comment (singaltanmay) [reply src/test/java/uk/ac/ebi/eva/contigalias/service/AssemblyAndChromosomeServiceIntegrationTest.java:187]: No specific detail. Just thought it would be too wordy and confusing to explain what is the "same" as to what. The reasoning behind these tests being separate is a bit confusing at first but I promise they test different cases.

File: src/main/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasController.java [hunk 5]
```diff
@@ -169,23 +169,58 @@ public ContigAliasController(ContigAliasHandler handler) {
...
+                    "return an HTTP error code 400 if invalid parameters are found.")
+    @GetMapping(value = "chromosomes/name/{name}/assembly/taxid/{taxid}")
```
[the remark this answers is not in the document]
Comment (singaltanmay) [reply src/main/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasController.java:189]: Do you want to change the URL? I could change it to `../chromosomes/assembly/taxid/{taxid}?name={name}`

File: src/main/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasController.java [hunk 6]
```diff
@@ -169,23 +169,58 @@ public ContigAliasController(ContigAliasHandler handler) {
...
+    }
 
+    @ApiOperation(value = "Get chromosomes using a combination of their own name and the GenBank or RefSeq accessions" +
+            " of their parent assemblies.",
+            notes = "Given a chromosome's name and the GenBank or RefSeq accession of the assembly that it belongs " +
+                    "to, this endpoint will return a list of chromosomes that satisfy the given parameters. Each " +
+                    "chromosome will also have it's parent assembly nested inside it. The endpoint will either return" +
+                    " a list of chromosomes or it will either return an HTTP error code 204 if no chromosomes are " +
+                    "found or return an HTTP error code 400 if invalid parameters are found.")
+    @GetMapping(value = "chromosomes/name/{name}/assembly/accession/{accession}")
```
Comment (singaltanmay) [inline src/main/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasController.java:211]: I think it would be a good idea to also change this to `chromosomes/{name}?accession={accession}`

File: src/main/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasController.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmmut) [reply src/main/java/uk/ac/ebi/eva/contigalias/controller/contigalias/ContigAliasController.java]: I have a similar issue with `assertEquals(expected, actual)` where I keep reversing the arguments, for years now.

File: src/test/java/uk/ac/ebi/eva/contigalias/service/AssemblyAndChromosomeServiceIntegrationTest.java (hunk 2 above)
Comment (jmmut) [reply src/test/java/uk/ac/ebi/eva/contigalias/service/AssemblyAndChromosomeServiceIntegrationTest.java:187]: yeah I understand the tests themselves, I was talking about the comments only.
